src/host-services.js

```diff
@@ -0,0 +1,150 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  getServicePromiseForDoc,
+  registerServiceBuilderForDoc,
+} from './service';
+
+const ServiceNames = {
+  VISIBILITY: 'host-visibility',
+  FULLSCREEN: 'host-fullscreen',
+  EXIT: 'host-exit',
+};
+
+export class HostServices {
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @return {!Promise<!Visibility>}
+   */
+  static visibilityForDoc(elementOrAmpDoc) {
+    return /** @type {!Promise<!Visibility>} */ (
+      getServicePromiseForDoc(elementOrAmpDoc, ServiceNames.VISIBILITY));
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @param {!Visibility} impl
+   */
+  static installVisibilityServiceForDoc(elementOrAmpDoc, impl) {
+    registerServiceBuilderForDoc(elementOrAmpDoc,
+        ServiceNames.VISIBILITY, impl, /* opt_instantiate */ true);
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @return {!Promise<!Visibility>}
+   */
+  static fullscreenForDoc(elementOrAmpDoc) {
+    return /** @type {!Promise<!Fullscreen>} */ (
+      getServicePromiseForDoc(elementOrAmpDoc, ServiceNames.FULLSCREEN));
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @param {!Fullscreen} impl
+   */
+  static installFullscreenServiceForDoc(elementOrAmpDoc, impl) {
+    registerServiceBuilderForDoc(elementOrAmpDoc,
+        ServiceNames.FULLSCREEN, impl, /* opt_instantiate */ true);
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @return {!Promise<!Exit>}
+   */
+  static exitForDoc(elementOrAmpDoc) {
+    return /** @type {!Promise<!Fullscreen>} */ (
+      getServicePromiseForDoc(elementOrAmpDoc, ServiceNames.EXIT));
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @param {!Exit} impl
+   */
+  static installExitServiceForDoc(elementOrAmpDoc, impl) {
+    registerServiceBuilderForDoc(elementOrAmpDoc,
+        ServiceNames.EXIT, impl, /* opt_instantiate */ true);
+  }
+}
+
+/**
+ * Visibility defines interface provided by host for visibility detection.
+ *
+ * @interface
+ */
+export class Visibility {
+
+  /**
+   * Register a callback for scroll events.
+   * @param {function(!VisibilityDataDef)} unusedCallback
+   */
+  onVisibilityChange(unusedCallback) {}
+}
+
+/**
+ * The structure that combines position and size for an element. The exact
+ * interpretation of position and size depends on the use case.
+ *
+ * @typedef {{
+ *   visibleRect: (?./layout-rect.LayoutRectDef),
+ *   visibleRatio: number
+ * }}
+ */
+export let VisibilityDataDef;
```

Comment:
What's the visibleRect here? 
Is it enough for scroll bound animation?

Reply:
we will need a different service for scroll bound animation. 

@jeffkaufman is scroll a thing in MRAID?